radicle-httpd/src/raw.rs

```diff
@@ -27,12 +29,16 @@ async fn file_handler(
 ) -> impl IntoResponse {
     let storage = &profile.storage;
     let repo = Repository::open(paths::repository(storage, &project))?;
-    let blob = repo.blob(sha, &path)?;
-
     let mut response_headers = HeaderMap::new();
+
+    if repo.file(sha, &path)?.content(&repo)?.size() > MAX_BLOB_SIZE {
```

Comment:
wasn't able to find in the source whether calling `.content(..)` allocates memory for the content itself (which would defeat the purpose here).

I suppose it's just a light pointer, but need your confirmation on this @FintanH

Reply:
I think it must load it in memory to know its size, since it's not necessarily a file. So probably best to only load it once.

Reply:
if it's loaded in memory, can't we just get metadata of blob from disk using `fs::`?

> So probably best to only load it once.

wdym? you want it cached?

Reply:
to give you a corner case: image there's a 1 GB binary on server and someone tries to download that, this would lead to OOM error on server, regardless of how many times we try to download it. 

now think of a smaller size binary, it can still be used to DDoS server, just increase the number of concurrent requests.

there's really no point in allocating but then not serving based on size, because serving part is not expensive, allocating is.

Reply:
if it ends up loading the content in order to know its size, then I suggest we get metadata of blob from disk using `fs::`.

to give you a corner case: image there's a 1 GB binary on server and someone tries to download that, this would lead to OOM error on a server (with 1GB of ram), regardless of how many times we load it.

now think of a smaller size binary, it can still be used to DDoS server, just increase the number of concurrent requests.

there's really no point in allocating but then not serving based on size, because serving part is not expensive, allocating is.

Reply:
We can't get the metadata of the blob from disk. A blob is not necessarily an on-disk file.

Reply:
> to give you a corner case: image there's a 1 GB binary on server and someone tries to download that, this would lead to OOM error on a server (with 1GB of ram), regardless of how many times we load it.

This is best handled by just not hosting repos with 1 GB binaries in them, since you'll have the same problem at the p2p layer.

Reply:
Digging into what is called by `Blob::size`, we can see it's these two calls [blob.c](https://github.com/libgit2/libgit2/blob/80742e159c3221fd853d9a0fe735b382a7e3419d/src/libgit2/blob.c#L29-L37) & [odb.c](https://github.com/libgit2/libgit2/blob/6f2eedb7eb58f5af2e96cafeae7261c07c928d89/src/libgit2/odb.c#L183-L186). It _seems_ like it doesn't need to load the data into memory to determine the size, so this should be fine.

Reply:
But we're doing `content().size()` above..?

Reply:
That actually turns the `Oid` in the `File` into the corresponding `git2::Blob`, which is just a pointer object.

Reply:
`git2::Blob` loads the blob from disk though.

Reply:
I don't _think_ it does judging by the [struct internals](https://docs.rs/git2/0.16.1/src/git2/blob.rs.html#12-15). Why do you think it does?

Reply:
The underlying `git_blob` type has a data pointer: https://github.com/libgit2/libgit2/blob/8a871d13b7f4e186b8ad943ae5a7fcf30be52e67/src/libgit2/blob.h#L17

Reply:
Hmmm it _could_ be the odb pointer instead https://github.com/libgit2/libgit2/blob/8a871d13b7f4e186b8ad943ae5a7fcf30be52e67/src/libgit2/blob.h#L21 since it's a `union`. I'm not adept enough at C to track when it actually changes to the data pointer though

Reply:
Yeah could question... I don't know in which cases the odb pointer is used, and how that all works.. The only way it would *not* have to preload the blob in memory is if the odb is mmaped I guess..